### PR TITLE
#1172 Default CacheKeyGenerator

### PR DESCRIPTION
**Pull request by raman-m**
## Follows up #1172 
- #1172 

## Related to #1808 
- #1808 

During [22.0](https://github.com/ThreeMammals/Ocelot/releases/tag/22.0.0) release feature #1172 was excluded from delivery list. Now it is time to deliver it once again with some enhancements after team's discussion ([start](https://github.com/ThreeMammals/Ocelot/pull/1172#issuecomment-1828721075)).

## Proposed Changes
  - Disable request content hashing for performance reasons
  - Introduce new `EnableRequestBodyHashing` property of the `CacheOptions` class
  - Make `CacheKeyGenerator` default via class renaming

### CODE REVIEW

**Comment by raman-m** [comment]
@ggnaegi You are welcome to review code of this PR!

We have some dev plan on this:
- your comments: [1st](https://github.com/ThreeMammals/Ocelot/pull/1172#issuecomment-1830081902) and [2nd](https://github.com/ThreeMammals/Ocelot/pull/1172#issuecomment-1830740495)
- after merging this PR you'll create new PR on top of this code

Please review!

**Comment by raman-m** [comment]
Should I add acceptance tests? 😇 😃

**Review by ggnaegi — CHANGES_REQUESTED** [review]
@raman-m a few changes i would recommend. Please no virtual, since we don't have any use cases with overridden properties and methods. The only use case here is testing...

**Comment by ggnaegi** [comment]
> Should I add acceptance tests? : 😇 😋

😇

File: src/Ocelot/Request/Middleware/DownstreamRequest.cs [hunk 1]
```diff
@@ -17,14 +19,13 @@ public DownstreamRequest(HttpRequestMessage request)
             Headers = _request.Headers;
             AbsolutePath = _request.RequestUri.AbsolutePath;
             Query = _request.RequestUri.Query;
-            Content = _request.Content;
         }
 
-        public HttpRequestHeaders Headers { get; }
+        public virtual HttpHeaders Headers { get; }
```
[the remark this answers is not in the document]
Comment (RaynaldM) [reply src/Ocelot/Request/Middleware/DownstreamRequest.cs:24]: agree with you @ggnaegi

**Review by RaynaldM — APPROVED** [review]

File: src/Ocelot/Cache/DefaultCacheKeyGenerator.cs [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+                    .Value?.FirstOrDefault();
+
+                if (!string.IsNullOrEmpty(header))
+                {
+                    builder.Append(Delimiter)
+                        .Append(header);
+                }
+            }
+
+            if (!downstreamRequest.HasContent || !options.EnableRequestBodyHashing)
```
[the remark this answers is not in the document]
Comment (raman-m) [reply src/Ocelot/Cache/DefaultCacheKeyGenerator.cs]: Fixed in https://github.com/ThreeMammals/Ocelot/pull/1849/commits/5305b306859f82dd08430a576d3ed12469adf820

File: src/Ocelot/Request/Middleware/DownstreamRequest.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (raman-m) [reply src/Ocelot/Request/Middleware/DownstreamRequest.cs]: Good recommendation!
Fixed in https://github.com/ThreeMammals/Ocelot/pull/1849/commits/faaed441064ecc81e309ffe09adc0a95d9c60da3 ✔️

File: src/Ocelot/Request/Middleware/DownstreamRequest.cs [hunk 3]
```diff
@@ -17,14 +19,13 @@ public DownstreamRequest(HttpRequestMessage request)
...
 
-        public string Method { get; }
+        public virtual string Method { get; }
 
-        public string OriginalString { get; }
+        public virtual string OriginalString { get; }
```
[the remark this answers is not in the document]
Comment (raman-m) [reply src/Ocelot/Request/Middleware/DownstreamRequest.cs]: Removed in https://github.com/ThreeMammals/Ocelot/pull/1849/commits/faaed441064ecc81e309ffe09adc0a95d9c60da3

File: src/Ocelot/Request/Middleware/DownstreamRequest.cs [hunk 4]
```diff
@@ -36,7 +37,9 @@ public DownstreamRequest(HttpRequestMessage request)
 
         public string Query { get; set; }
 
-        public HttpContent Content { get; set; }
+        public virtual bool HasContent { get => _request?.Content != null; }
```
[the remark this answers is not in the document]
Comment (raman-m) [reply src/Ocelot/Request/Middleware/DownstreamRequest.cs]: Removed in https://github.com/ThreeMammals/Ocelot/pull/1849/commits/faaed441064ecc81e309ffe09adc0a95d9c60da3

File: src/Ocelot/Request/Middleware/DownstreamRequest.cs [hunk 5]
```diff
@@ -36,7 +37,9 @@ public DownstreamRequest(HttpRequestMessage request)
...
+
+        public virtual HttpRequestMessage Request { get => _request; }
```
[the remark this answers is not in the document]
Comment (raman-m) [reply src/Ocelot/Request/Middleware/DownstreamRequest.cs]: Removed in https://github.com/ThreeMammals/Ocelot/pull/1849/commits/faaed441064ecc81e309ffe09adc0a95d9c60da3

File: test/Ocelot.UnitTests/Cache/DefaultCacheKeyGeneratorTests.cs [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+using System.Net.Http.Headers;
+using System.Reflection;
+using System.Text;
+
+namespace Ocelot.UnitTests.Cache
+{
+    public class DefaultCacheKeyGeneratorTests
+    {
+        private readonly ICacheKeyGenerator _cacheKeyGenerator;
+        private readonly Mock<DownstreamRequest> _downstreamRequest;
```
[the remark this answers is not in the document]
Comment (raman-m) [reply test/Ocelot.UnitTests/Cache/DefaultCacheKeyGeneratorTests.cs]: Good recommendation!
Fixed in https://github.com/ThreeMammals/Ocelot/pull/1849/commits/faaed441064ecc81e309ffe09adc0a95d9c60da3 ✔️

File: test/Ocelot.UnitTests/Cache/DefaultCacheKeyGeneratorTests.cs [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+            _downstreamRoute = new DownstreamRouteBuilder()
+                .WithKey("key1")
+                .WithCacheOptions(options)
+                .Build();
+        }
+
+        private string _generatedCacheKey;
+
+        private async Task WhenGenerateRequestCacheKey()
+        {
```
[the remark this answers is not in the document]
Comment (raman-m) [reply test/Ocelot.UnitTests/Cache/DefaultCacheKeyGeneratorTests.cs:103]: Done by https://github.com/ThreeMammals/Ocelot/pull/1849/commits/faaed441064ecc81e309ffe09adc0a95d9c60da3 ✔️

File: test/Ocelot.UnitTests/Cache/DefaultCacheKeyGeneratorTests.cs [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+            _generatedCacheKey = await _cacheKeyGenerator.GenerateRequestCacheKey(_downstreamRequest.Object, _downstreamRoute);
+        }
+
+        private void ThenGeneratedCacheKeyIs(string expected)
+        {
+            _generatedCacheKey.ShouldBe(expected);
+        }
+    }
+
+    internal class HttpHeadersStub : HttpHeaders
```
[the remark this answers is not in the document]
Comment (raman-m) [reply test/Ocelot.UnitTests/Cache/DefaultCacheKeyGeneratorTests.cs]: Stub class has been removed in https://github.com/ThreeMammals/Ocelot/pull/1849/commits/faaed441064ecc81e309ffe09adc0a95d9c60da3 ✔️

File: test/Ocelot.UnitTests/Cache/DefaultCacheKeyGeneratorTests.cs [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+            this.Given(x => x.GivenDownstreamRoute(options))
+                .When(x => x.WhenGenerateRequestCacheKey())
+                .Then(x => x.ThenGeneratedCacheKeyIs(cachekey))
+                .BDDfy();
+        }
+
+        [Fact]
+        public void should_generate_cache_key_happy_path()
+        {
+            const string content = nameof(should_generate_cache_key_happy_path);
```
[the remark this answers is not in the document]
Comment (raman-m) [reply test/Ocelot.UnitTests/Cache/DefaultCacheKeyGeneratorTests.cs:74]: Added new `GivenHasContent` helper in https://github.com/ThreeMammals/Ocelot/pull/1849/commits/faaed441064ecc81e309ffe09adc0a95d9c60da3 ✔️

**Comment by raman-m** [comment]
>@ggnaegi suggested changes [on Dec 9](https://github.com/ThreeMammals/Ocelot/pull/1849#pullrequestreview-1773723485)
>@RaynaldM approved these changes [on Dec 11](https://github.com/ThreeMammals/Ocelot/pull/1849#pullrequestreview-1774415198)

Code review issues have been **fixed** ✅

**Comment by raman-m** [comment]
> @ggnaegi commented [on Dec 9](https://github.com/ThreeMammals/Ocelot/pull/1849#issuecomment-1848584274)

Acceptance test has been [added](https://github.com/ThreeMammals/Ocelot/blob/follow-up-1172/test/Ocelot.AcceptanceTests/Caching/CachingTests.cs#L116-L153)! ✔️ 

By default body content is not hashed.
@ggnaegi  Looks good now?

**Review by ggnaegi — APPROVED** [review]

**Comment by raman-m** [comment]
Ready for delivery! ✔️

File: src/Ocelot/Configuration/CacheOptions.cs [hunk 10]
```diff
@@ -1,15 +1,39 @@
...
+        /// <summary>
+        /// Enables MD5 hash calculation of the <see cref="HttpRequestMessage.Content"/> of the <see cref="DownstreamRequest.Request"/> object.
+        /// </summary>
+        /// <remarks>
+        /// Default value is <see langword="false"/>. No hashing by default.
+        /// </remarks>
+        /// <value>
+        /// <see langword="true"/> if hashing is enabled, otherwise it is <see langword="false"/>.
+        /// </value>
+        public bool EnableContentHashing { get; }
```
Comment (wast) [inline src/Ocelot/Configuration/CacheOptions.cs:37]: I think this was breaking change that causes wrong caches with POST requests with payload. Rolled back to version 21 where there is no issue. Haven't had time to confirm, but I'm 90% sure. Gonna try to add a unit test in a PR.

**Comment by raman-m** [comment]
Dear @wast, don't remove your comments: better to discuss what was happen. Anyway I have email notifications.

> ```diff
> +        public bool EnableContentHashing { get; }
> ```
> I think this was breaking change that causes wrong caches with POST requests with payload. Rolled back to version 21 where there is no issue. Haven't had time to confirm, but I'm 90% sure. Gonna try to add a unit test in a PR.
> 

The problem was reported in #2054 #2059, and the bug was fixed by #2058 with merging on May 13, 2024, see commit https://github.com/ThreeMammals/Ocelot/commit/6e9a975fae3b26229c3f0b3c1a475a7b05e95fee please. So, the fix was delivered in version [23.3.0](https://github.com/ThreeMammals/Ocelot/releases/tag/23.3.0).

> Gonna try to add a unit test in a PR.

What is the purpose? The bug was fixed by #2058! Therefore, please use the latest version and specify [global options](https://ocelot.readthedocs.io/en/latest/features/caching.html#global-configuration). Enjoy!

**Comment by wast** [comment]
Comment was deleted as it wasn't relevant anymore.

**Pull request merged**